### ISSUE TITLE
Make type projections safe instead of dropping them

### ISSUE THREAD

**Issue by LPTK** [lampepfl/dotty#5998]
## Background

It _seems_ that type projections, which are of the form `T # X`, could be made sound if we simply disregarded the lower bound of `X` in `T` unless `T` is a singleton type.

For example:
```scala
class T { type X >: A <: B }
infer[T # X <:< B] // should work
infer[T # A <:< X] // should NOT work
```

See also [the discussions](https://contributors.scala-lang.org/t/proposal-to-remove-general-type-projection-from-the-language/2812/76) on the contributors forum.

(Note: we could just as well disregard the upper bound instead, but I suppose that would be much less useful.)

## Pratical Motivation

Type projections are used quite a lot in Scala 2. It would be best if we could just port most of that code as is, since migrating to not using type projections requires extensive refactoring, cannot be automated, and may introduce performance concerns (see [the same forum discussion](https://contributors.scala-lang.org/t/proposal-to-remove-general-type-projection-from-the-language/2812)).

The thing is, it seems it would be possible to port most of that existing Scala 2 code to a system with the restriction I proposed above.

I have seen lots of code where (non-trivial) upper bounds of type projections were used. I don't remember seeing lots of code where lower bounds were used. And I'm pretty sure I've never seen code where both bounds of a type projections were used, except in constructions meant to show the unsoundness of type projections.

## Philosophical Motivation

Match types are a great addition to the language, in that they give us a concise and lightweight way of expressing _closed_ type families. With https://github.com/lampepfl/dotty/pull/5996 they should even become robust in the face of abstract types.

Dually, type projections have been great because they basically give us concise and lightweight syntax for expressing _open_ type families. If we add closed type families to Scala, it makes sense to me that we should strive to _continue_ supporting open ones!

## Semantics

What does a type projection `T # X` mean? I would say it means _some type_ with a known upper bound, on which we know nothing more... that is, until `T` is substituted with a type in which `X` is concrete (at which point we can just pick that type out), or substituted with a singleton type (at which point we have a classical path-dependent type). It's easy to see that in both situations, such substitution will satisfy the upper bound that we have assumed for `T # X`.

Is that kind of weird? Well, IMHO it's no weirder than the nature of a non-reduced match type: a type on which we just know an upper bound (the union of the cases), but that is only made to precisely _mean_ something when the scrutinee becomes concrete enough that the match stops being ambiguous.

---

Have I missed something? Would that actually be unsound?

**Comment by odersky** [comment]
We'd need a worked out way to map this down to DOT. Otherwise there's no way to answer the question:

> Have I missed something? Would that actually be unsound?

I believe it would be great if someone undertook this effort. But foundations have to come first, and they are not optional.

**Comment by LPTK** [comment]
I agree on the principle. But AFAIK several current features of Dotty do not yet have DOT mappings either, such as higher-kinded types _and match types_.

Both higher-kinded types and (restricted) type projections are important for backward compatibility with Scala 2. I think that should justify their inclusion even though they have not been formally proven correct yet. At least we have a good intuition that they are "morally" sound (but of course, that intuition may turn out to be wrong).

**Comment by Blaisorblade** [comment]
@LPTK Somebody must make a judgment call for unproven addition, based on what we know anyway. Unlike for higher kinds, we know type projections have caused big trouble before.

I'm thinking again of how to add type projections to my proofs: one needs to define when values are in `T#X`, and check the type rules that follow. But using the semantics I and Miles suggested in #4583 doesn't seem to give the behavior you're asking for; one would require some twists.

==

Separately, re-reading #4583, it seems what you propose might be close to the original semantics for type projection in the 2003 paper:
> Formally, the crucial rules in the nuObj paper are (Alias-=) in Sec. 4.2, (Tsel-\prec (strange <)) in 4.3, (Tsel-<:) in 4.4. They all refer to membership in (Other-\in) in Sec. 4.1, which does introduce an existential x: T. Since there are no lower bounds, those rules can only show T#A <: U (in case A's decl is type A >: L <: U) or T#A = Z (in case A's decl is type A = Z), but never T#A >: L.
However, I'm not sure if that paper gives good answers on intersections of conflicting bounds (and I doubt it).

**Comment by LPTK** [comment]
@Blaisorblade by "using the semantics I and Miles suggested", are you referring to this?

> By `t.A forAll { val t: T }` I intended the _intersection_ of the `A`s for all `t: T`, not the union

How could that be the intended semantics? It would mean that given _any_ `a:A`, you can upcast `A#T` to `a.T`, and that you _could not_ upcast `a.T` to `A#T`.
This clearly goes against the intended semantics of type projections in the particular case of nested classes, which for the record I think should be the following:
```scala
class A { self => class B { /* defs */ } }
```
is to be understood as:
```scala
class A$B(self: A) { /* defs */ }
class A { type B <: A$B }
```
where `A#B` _denotes_ that common ancestor class `A$B` of all instances of the inner class`B` irrelevant of which `A` they were instantiated from. In that sense, a type projection `A#T` would be the _union_ of the types `a.T` living in all possible instances `a` of the prefix `A`.

**Comment by Blaisorblade** [comment]
> @Blaisorblade by "using the semantics I and Miles suggested", are you referring to this?
> 
> > By `t.A forAll { val t: T }` I intended the _intersection_ of the `A`s for all `t: T`, not the union
> 
> How could that be the intended semantics? It would mean that given _any_ `a:A`, you can upcast `A#T` to `a.T`, and that you _could not_ upcast `a.T` to `A#T`.

As I mentioned to @LPTK elsewhere. For the record, while Miles and I might not yet agree and he might be right, I favor using the union.

Assigning this to myself, as I'll keep this among my TODOs in my theoretical work.
